lib/charms/hpc_libs/v0/slurm_ops.py

```diff
@@ -197,9 +148,183 @@ def config_name(self) -> str:
+        ...
+
+    @abstractmethod
+    def set(self, config: Mapping[str, Any]) -> None:
+        """Set configuration for Slurm component."""
+        ...
+
+    @abstractmethod
+    def unset(self, *keys: str) -> None:
+        """Unset configuration for Slurm component."""
+        ...
+
+
+class MungekeyManager(ABC):
+    """Control the munge key."""
+
+    def get(self) -> str:
+        """Get the current munge key.
+
+        Returns:
+            The current munge key as a base64-encoded string.
+        """
+        ...
+
+    def set(self, key: str) -> None:
+        """Set a new munge key.
+
+        Args:
+            key: A new, base64-encoded munge key.
+        """
+        ...
+
+    def generate(self) -> None:
+        """Generate a new, cryptographically secure munge key."""
+        ...
+
+
+class SlurmPackageManager(ABC):
+    """Manager to control the installation, creation and configuration of Slurm-related services."""
+
+    @abstractmethod
+    def install(self) -> None:
+        """Install Slurm."""
+        ...
+
+    @abstractmethod
+    def version(self) -> str:
+        """Get the current version of Slurm installed on the system."""
+        ...
+
+    @abstractmethod
+    def service_for(self, service: ServiceType) -> ServiceManager:
+        """Get a `ServiceManager` for the specified `service`."""
+        ...
+
+    @abstractmethod
+    def configuration_manager_for(self, service: ServiceType) -> ConfigurationManager:
+        """Get a `ConfigurationManager` for the specified `service`."""
+        ...
+
+    @abstractmethod
+    def mungekey_manager(self) -> MungekeyManager:
+        """Get the `MungekeyManager` for munge."""
+        ...
+
+
+class MungeManager:
+    """Manage `munged` service operations."""
+
+    def __init__(self, ops_manager: SlurmPackageManager) -> None:
+        self.service = ops_manager.service_for(ServiceType.MUNGED)
+        self.config = ops_manager.configuration_manager_for(ServiceType.MUNGED)
+        self.key = ops_manager.mungekey_manager()
+
+
+class PrometheusExporterManager:
+    """Manage `slurm-prometheus-exporter` service operations."""
+
+    def __init__(self, ops_manager: SlurmPackageManager) -> None:
+        self.service = ops_manager.service_for(ServiceType.PROMETHEUS_EXPORTER)
+
+
+class SlurmManagerBase:
+    """Base manager for Slurm services."""
+
+    def __init__(self, service: ServiceType, ops_manager: SlurmPackageManager) -> None:
+        self.service = ops_manager.service_for(service)
+        self.config = ops_manager.configuration_manager_for(service)
+        self.munge = MungeManager(ops_manager)
+        self.exporter = PrometheusExporterManager(ops_manager)
+        self.ops = ops_manager
+
+    @property
+    def hostname(self) -> str:
+        """The hostname where this manager is running."""
+        return socket.gethostname().split(".")[0]
+
+
+class SnapPackageManager(SlurmPackageManager):
```

Comment:
Should this just be `SnapManager` instead? Reason saying is that it's also providing methods for accessing both the service and configuration managers. _Less characters to type the better_ :rofl:

```suggestion
class SnapManager(SlurmPackageManager):
```